core/shared/src/main/scala/zio/internal/FiberRuntime.scala

```diff
@@ -654,18 +670,32 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
   private def interruptAllChildren(): UIO[Any] =
     if (sendInterruptSignalToAllChildren(_children)) {
       val iterator = _children.iterator()
-
       _children = null
 
-      val body = () => {
-        val next = iterator.next()
+      var curr: Fiber.Runtime[_, _] = null
 
-        if (next != null) next.await(id.location) else Exit.unit
+      def skip() = {
+        curr = null
+        while (iterator.hasNext && (curr eq null)) {
+          curr = iterator.next()
+          if ((curr ne null) && !curr.isAlive())
+            curr = null
+        }
       }
 
-      // Now await all children to finish:
-      ZIO
-        .whileLoop(iterator.hasNext)(body())(_ => ())(id.location)
+      skip()
+
+      if (null ne curr) {
+        val body = () => {
+          val c = curr
+          skip()
+          c.await(id.location)
+        }
+
+        // Now await all children to finish:
+        ZIO
+          .whileLoop(null ne curr)(body())(_ => ())(id.location)
+      } else null
```

Comment:
I was struggling to follow the logic here. I think the previous code could have been minorly adjusted to make it more performant as:

```scala
      val body = () => {
        var next = null
        while (next eq null && iterator.hasNext) {
          val next0 = iterator.next()
          if (next0 ne null && next0.isAlive()) next = next0
        }

        if (next ne null) next.await(id.location) else Exit.unit
      }
```

What do you think?

Reply:
might work, I'll have a deeper look at it.
the basic idea is to efficiently implement a kind of filtering iterator here without actually paying for one (scala's filtering iterator is effectively a buffered iterator which introduces some costs to its `hasNext` implementation)

Reply:
@kyri-petrou I was actually able to drop `body` altogether and reduce the `whileLoop` to this:

```
ZIO
          .whileLoop(null ne curr)(curr.await(id.location))(_ => skip())(id.location)
```

in fact, there's another possible additional optimization here:
by the time we return the `whileLoop` effect we already have the first child fiber, so we can separate the first iteration from the rest by doing something like: 
```ZIO.suspend(val effect = curr.await; skip(); effect) *> whileLoop(...)```

it'd also be nice if there was a `doWhile` effect...

Reply:
This is a Java iterator though (`_children` is a Java Set).

In either way, do we know how many of the children here are expected to be still alive? What use-case do you have where the children are alive by the time the fiber exits? Does this relate to anything in this PR, and if yes, are we better off waiting for it to be merged first?

https://github.com/zio/zio/pull/8816

Reply:
> ```
> ZIO
>           .whileLoop(null ne curr)(curr.await(id.location))(_ => skip())(id.location)
> ```

Ah nice, I like this. Just maybe write it as `curr ne null` cause it's a bit uncommon the other way around 😅

> in fact, there's another possible additional optimization here: by the time we return the `whileLoop` effect we already have the first child fiber, so we can separate the first iteration from the rest by doing something like: `ZIO.suspend(val effect = curr.await; skip(); effect) *> whileLoop(...)`

I think that's a _very micro_ micro-optimization at the cost of readability. I don't think we're saving anything more here other than 1 `curr ne null` check as far as I can tell

Reply:
@kyri-petrou I don't think this is related to #8816 , furthermore when a fiber terminates all of its child fibers (and grand children and all descendants in general) are interrupted by definition (structured concurrency and such...) so I'm not 100% why `ZIO.foreachPar` has to interfere with this behavior in any way... but again irrelevant to this PR.

regarding the two operations optimized by this PR, in the happy case I expect exactly zero running child fibers when the stream completes, hence there should be no fibers to interrupt when the parent fiber completes, this is of course not 100% true... streams may fail, for the mapZIOParXXX operators this means some operations may still be running and have to be interrupted. there are also cases where the stream itself doesn't fail but downstream cancels by applying an operator like `takeN`.

so summing this up, in the context of these two stream operators and probably for most correct fibers aware code I'd expect no pending fibers or really just a few on successful completion, in case of failure/interruption there might be more and some complicated code may actually spawn children expecting them to be terminated on completion.

so the motivation behind this micro(scopic) optimization was the expectation of very few pending child fibers.

Reply:
@kyri-petrou once a career starts with ~15 years of writing C++, you put the nulls first on equality checks 😎

Reply:
Regarding https://github.com/zio/zio/pull/8816, I agree with you, `ZIO.foreachPar` shouldn't be interfering. But there is a use-case (imho not a valid one) that would be affected if we changed that behaviour). Check the PR description for more info on that